### ISSUE TITLE
Create a vault factory

### ISSUE THREAD

**Issue by ASlowCheetah** [pangolin-defi/exchange-contracts#2]
**Acceptance Criteria:**
* Anyone can use it to create new vault
* Anyone can list all existing vaults
* Anyone can list all existing vaults that contain exact list of provided tokens
* Tokens order in the vaults needs to be consistent, suggested ordering is to use the token address as comparison

Note:
* It's a good idea to have a look on how PangolinFactory does to instantiate new contracts, there's a bit of low level assembly going on there. The generation of the address needs to be unvarying such that the same set of tokens and amounts will always generate the same deployed address.

**Comment by benber86** [comment]
> The generation of the address needs to be unvarying such that the same set of tokens and amounts will always generate the same deployed address.

Do we really want to do this?
Say somebody starts a 60/40 BTC/ETH fund with a 2% management fee. I also want to start a 60/40 fund, but do the management myself. Why should I be prevented from creating such a fund just because someone started it first? Same issue if someone creates the fund and turns out to be incompetent at managing it, it's useful to have the ability to create competing funds with different distributions.

This also assumes that the token sets associated with each vault are immutable. But we want to stay flexible there. Say someone starts a WBTC/WETH fund, but the address of WETH on Avalanche changes. There needs to be a migration. Or someone has a DeFi fund, but one of the project fails after a couple of years and the token needs to be removed from the fund's portfolio.

> Anyone can list all existing vaults that contain exact list of provided tokens

Why not just have a list of all vaults and all vaults will have getter methods to get their token sets + associated weights. Keeping a copy in the factory contract seems redundant.  It assumes vault immutability (or otherwise we have to update factory + vault everytime we update the vault). 

I understand the appeal from a front-end perspective, but the dapp's front can either make multiple calls to vault contracts or keep an up to date index on a separate backend.

**Comment by ASlowCheetah** [comment]
>  Say somebody starts a 60/40 BTC/ETH fund with a 2% management fee. I also want to start a 60/40 fund, but do the management myself. Why should I be prevented from creating such a fund just because someone started it first? Same issue if someone creates the fund and turns out to be incompetent at managing it, it's useful to have the ability to create competing funds with different distributions.

That's true, lets follow with incremental address, and let the VaultFactory store all the possible vault addresses for a given token set

So it needs additional function to allow one vault to migrate from a certain token set register, to another, in case of a balancing happens and one token is removed/added

> Why not just have a list of all vaults and all vaults will have getter methods to get their token sets + associated weights. Keeping a copy in the factory contract seems redundant. It assumes vault immutability (or otherwise we have to update factory + vault everytime we update the vault).

My worry here is to have a consistent way to identify and search for vaults from the platform in a fast way, pooling from the contract constantly to find what you need might be slow on the long run.

The appeal is not only for the front-end, other dApps my want to build on top of our contracts, providing a feasible option for onchain search can be helpful, but I agree this might be more like "Nice to have" than a "Must have"
